Add matching against external people

chrismytton opened the pull request:
This adds a new model/DB table and three management commands.

The `ExternalPerson` model is for importing lists of known identifiers into along with the flag that they apply to.

Management commands:

- `generate_fake_popolo` uses the `BodsPersonStatement` model to generate fake popolo data for 10 random people for the next step. (Redirect it to a JSON file in the shell to save it.)
- `load_identifiers_from_popolo` takes a Popolo file (e.g. one created with the `generate_fake_popolo` command) and creates `ExternalPerson` objects from them
- `scan_contracts` - This has been expanded to go through the `BODSPersonStatement` identifiers and looks for matching `ExternalPerson` rows. If it finds one it creates a `FlagAttachment` so that person will be flagged.

## Notes

~~`KnownIdentifier` model feels a bit odd at the moment because it's very similar to the `FlagAttachment` model. I think the key difference is that `KnownIdentifer` represents _external_ identifiers and `FlagAttachment` represents an internal connection between `Flag` and other entities.~~ Fixed by renaming to `ExternalPerson`.

I _think_ there is value in this separation, because it allows us to expand the known identifiers to incorporate other information (perhaps known identifiers isn't great name, thinking about it 🤔 ) such as names, date of birth, address etc for matching on.

The main difficulty I'm having is knowing what kind of external data we're expecting to deal with, so we know how we might match it to internal data. At the moment generating fake data makes this all a bit sterile and hand-wavy.

SimKennedy commented:
Ok @chrismytton , I think I’ve got my head around the process here.
My understanding of how Bluetail would work _in the real world_, is that a procurement team would load in their OCDS data, then lookup and grab all the BODS data for all the referenced companies/people, then run a set of checks against all this that would highlight (flag) anything that the procurement officer should know. 
My thoughts are that these “checks” are independent scripts that look for specific issues and then create the FlagAttachment objects. For example, one check might be:
  
- Are any of the BODS entity/persons also in _this_ list of suspicious people/companies? If so let’s add a FlagAttachment for their identifier.

In that sense, your approach looks good, but as you’ve said, it seems the KnownIdentifier model is slightly duplicating the Flag Attachment purpose by also having a ForeignKey to Flag. Whereas it could just be the "naughty" list that we use to create the FlagAttachments?

Personally, I feel it’s slightly confusing (and redundant) trying to write scripts that fake the Flags, as opposed to writing scripts that do actual checks and Flagging as it would in the real world, but fake the reasons for the Flag. ie. We create the fake lists of suspicious people and fake tenders that have problems, but the “checks” themselves are valid and usable in the real world?

I realise this might be more effort though and so your approach of storing a list of fake flags and loading them in is a faster approach. But to avoid confusion, I think we need to be really clear about the model and management command naming and the purpose of each.

chrismytton commented:
@SimKennedy @crowbot Thanks both for your feedback!

I've made the following changes based on your comments:

- Renamed the `KnownIdentifier` model to `ExternalPerson`
- Moved the logic from the old `flag_contracts` command into the `scan_contracts` command
- Added an `ExternalPerson.match method`, which is where we can potentially hook in matching other things
- Updated `script/setup` to run these new management commands

SimKennedy commented on README.md:
```diff
@@ -222,6 +222,35 @@ There is a management command, `scan_contracts`, which looks through all
+identifiers of people and companies in contracts with `ExternalPerson` records.
+When it finds a match it creates a `FlagAttachment` record, which uses the
+internal identifier from the bods data to attach the flag.
+
```
Perhaps it would be helpful to also have here a quick step by step guide of the process and commands to run to import an external list and create flags against it? And also the fake implementation as an example?

SimKennedy commented:
A couple of changes made to fix the Flagging on the tenders list page:
- The `scan_contracts` command executes another function `self.flag_tenders_linked_to_external_people()` to find the tenders related to the flagged external people and create FlagAttachments including the OCID
- This function also updated to ONLY create a single FlagAttachment per person per OCID using the OpenOwnership identifier, to avoid double counting when multiple identifers exist for a BODS Person

chrismytton commented:
Thanks for adding those fixes, @SimKennedy! Happy for you to merge this when you're ready 👍
